File: backend/internal/handlers/files.go
package handlers

import (
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/novapanel/novapanel/internal/models"
)

type FileHandler struct{}

func NewFileHandler() *FileHandler {
	return &FileHandler{}
}

func sanitizePath(base, requested string) (string, error) {
	clean := filepath.Clean(requested)
	full := filepath.Join(base, clean)
	if !strings.HasPrefix(full, base) {
		return "", fmt.Errorf("path traversal attempt blocked")
	}
	return full, nil
}

// GET /api/v1/files?path=/var/www
func (h *FileHandler) List(c *gin.Context) {
	reqPath := c.DefaultQuery("path", "/var/www")
	base := "/var/www"
	fullPath, err := sanitizePath(base, reqPath)
	if err != nil {
		c.JSON(http.StatusForbidden, models.ErrorResponse{Error: err.Error()})
		return
	}

	entries, err := os.ReadDir(fullPath)
	if err != nil {
		c.JSON(http.StatusNotFound, models.ErrorResponse{Error: "Directory not found"})
		return
	}

	var files []models.FileEntry
	for _, entry := range entries {
		info, _ := entry.Info()
		fe := models.FileEntry{
			Name:  entry.Name(),
			Path:  filepath.Join(reqPath, entry.Name()),
			IsDir: entry.IsDir(),
		}
		if info != nil {
			fe.Size = info.Size()
			fe.Permissions = info.Mode().String()
			fe.ModifiedAt = info.ModTime().Format("2006-01-02 15:04:05")
		}
		files = append(files, fe)
	}

	c.JSON(http.StatusOK, files)
}

// GET /api/v1/files/content?path=/var/www/index.html
func (h *FileHandler) Read(c *gin.Context) {
	reqPath := c.Query("path")
	if reqPath == "" {
		c.JSON(http.StatusBadRequest, models.ErrorResponse{Error: "path is required"})
		return
	}

	base := "/var/www"
	fullPath, err := sanitizePath(base, reqPath)
	if err != nil {
		c.JSON(http.StatusForbidden, models.ErrorResponse{Error: err.Error()})
		return
	}

	info, err := os.Stat(fullPath)
	if err != nil {
		c.JSON(http.StatusNotFound, models.ErrorResponse{Error: "File not found"})
		return
	}

	if info.Size() > 5*1024*1024 {
		c.JSON(http.StatusBadRequest, models.ErrorResponse{Error: "File too large (> 5MB)"})
		return
	}

	content, err := os.ReadFile(fullPath)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{Error: "Failed to read file"})
		return
	}

	c.JSON(http.StatusOK, models.FileContentResponse{
		Path:    reqPath,
		Content: string(content),
		Size:    info.Size(),
	})
}

// PUT /api/v1/files/content
func (h *FileHandler) Write(c *gin.Context) {
	var req struct {
		Path    string `json:"path" binding:"required"`
		Content string `json:"content"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, models.ErrorResponse{Error: err.Error()})
		return
	}

	base := "/var/www"
	fullPath, err := sanitizePath(base, req.Path)
	if err != nil {
		c.JSON(http.StatusForbidden, models.ErrorResponse{Error: err.Error()})
		return
	}

	if err := os.MkdirAll(filepath.Dir(fullPath), 0755); err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{Error: "Failed to create directory"})
		return
	}

	if err := os.WriteFile(fullPath, []byte(req.Content), 0644); err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{Error: "Failed to write file"})
		return
	}

	c.JSON(http.StatusOK, models.SuccessResponse{Message: "File saved"})
}

// POST /api/v1/files/upload
func (h *FileHandler) Upload(c *gin.Context) {
	targetDir := c.PostForm("path")
	if targetDir == "" {
		targetDir = "/var/www"
	}

	base := "/var/www"
	fullDir, err := sanitizePath(base, targetDir)
	if err != nil {
		c.JSON(http.StatusForbidden, models.ErrorResponse{Error: err.Error()})
		return
	}

	file, header, err := c.Request.FormFile("file")
	if err != nil {
		c.JSON(http.StatusBadRequest, models.ErrorResponse{Error: "No file provided"})
		return
	}
	defer file.Close()

	if err := os.MkdirAll(fullDir, 0755); err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{Error: "Failed to create directory"})
		return
	}

	destPath := filepath.Join(fullDir, header.Filename)
	out, err := os.Create(destPath)
	if err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{Error: "Failed to create file"})
		return
	}
	defer out.Close()

	if _, err := io.Copy(out, file); err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{Error: "Failed to write upload"})
		return
	}

	c.JSON(http.StatusOK, models.SuccessResponse{Message: fmt.Sprintf("Uploaded %s", header.Filename)})
}

// DELETE /api/v1/files?path=/var/www/old.html
func (h *FileHandler) Delete(c *gin.Context) {
	reqPath := c.Query("path")
	if reqPath == "" {
		c.JSON(http.StatusBadRequest, models.ErrorResponse{Error: "path is required"})
		return
	}

	base := "/var/www"
	fullPath, err := sanitizePath(base, reqPath)
	if err != nil {
		c.JSON(http.StatusForbidden, models.ErrorResponse{Error: err.Error()})
		return
	}

	if err := os.RemoveAll(fullPath); err != nil {
		c.JSON(http.StatusInternalServerError, models.ErrorResponse{Error: "Failed to delete"})
		return
	}

	c.JSON(http.StatusOK, models.SuccessResponse{Message: "Deleted successfully"})
}

File: frontend/src/services/files.ts
import api from './api';

export const fileService = {
    async list(serverId: string, path = '/') {
        const { data } = await api.post('/files/list', { server_id: serverId, path });
        return data;
    },
    async read(serverId: string, path: string) {
        const { data } = await api.post('/files/read', { server_id: serverId, path });
        return data;
    },
    async write(serverId: string, path: string, content: string) {
        const { data } = await api.post('/files/write', { server_id: serverId, path, content });
        return data;
    },
    async create(serverId: string, path: string, isDir = false) {
        const { data } = await api.post('/files/create', { server_id: serverId, path, is_dir: isDir });
        return data;
    },
    async rename(serverId: string, oldPath: string, newPath: string) {
        const { data } = await api.post('/files/rename', { server_id: serverId, old_path: oldPath, new_path: newPath });
        return data;
    },
    async copy(serverId: string, src: string, dest: string) {
        const { data } = await api.post('/files/copy', { server_id: serverId, src, dest });
        return data;
    },
    async move(serverId: string, src: string, dest: string) {
        const { data } = await api.post('/files/move', { server_id: serverId, src, dest });
        return data;
    },
    async remove(serverId: string, path: string) {
        const { data } = await api.post('/files/delete', { server_id: serverId, path });
        return data;
    },
    async chmod(serverId: string, path: string, mode: string) {
        const { data } = await api.post('/files/chmod', { server_id: serverId, path, mode });
        return data;
    },
    async chown(serverId: string, path: string, owner: string) {
        const { data } = await api.post('/files/chown', { server_id: serverId, path, owner });
        return data;
    },
    async search(serverId: string, path: string, query: string) {
        const { data } = await api.post('/files/search', { server_id: serverId, path, query });
        return data;
    },
    async info(serverId: string, path: string) {
        const { data } = await api.post('/files/info', { server_id: serverId, path });
        return data;
    },
    async extract(serverId: string, path: string, dest?: string) {
        const { data } = await api.post('/files/extract', { server_id: serverId, path, dest });
        return data;
    },
    async compress(serverId: string, path: string, dest?: string) {
        const { data } = await api.post('/files/compress', { server_id: serverId, path, dest });
        return data;
    },
    async grep(serverId: string, path: string, pattern: string) {
        const { data } = await api.post('/files/grep', { server_id: serverId, path, pattern });
        return data;
    },
};
